maven_iuvs/echelle.py

```diff
@@ -334,11 +334,163 @@ def downselect_data(index, light_dark=None, orbit=None, date=None, segment=None,
+    the user's local hard drive. 
+
+    Parameters
+    ----------
+    keyfile : string
+              .csv file containing columns 'Light Folder', 'Light', 'Dark  
+              Folder', 'Dark', 'DTobj, 'Segment'.
+    ech_l1a_idx : list of dictionaries
+                  Metadata of local files, used to determine newest filename.
+    dark_idx : list of dictionaries
+               Same as above but the dark files only.
+    verbose : bool
+              If True will print result messages.
+
+    Returns
+    ----------
+    MASTER_KEY : pandas DataFrame
+                 New master key. 10s of thouasnds of lines
+    changes : pandas DataFrame
+              Lists old files and what they were replaced with
+    problems : pandas DataFrame
+               Files that caused a problem
+    """
+    MASTER_KEY = pd.read_csv(keyfile)
+
+    changes = {"row": [], "oldname": [], "newname": []}
+    problems = {"row": [], "oldname": [], "newname": []}
+
+    # Precompile regex
+    versioned_pat = re.compile(r"^(?P<base>mvn.+?)_(r|s)(?P<rev>\d+)(?P<ext>\.fits\.gz)$")
+
+    # Cache for folder listings
+    light_folder_cache = {}
+    dark_folder_cache = {}
+
+    new_lightnames = MASTER_KEY["Light"].copy()
+    new_darknames = MASTER_KEY["Dark"].copy()
+
+    for i, row in tqdm(MASTER_KEY.iterrows(), total=len(MASTER_KEY)):
+
+        # Lights
+        lfolder = row["Light Folder"]
+        dfolder = row["Dark Folder"]
+        original_lightname = row["Light"]
+        original_darkname = row["Dark"]
+
+        # Cache file list in this folder
+        if lfolder not in light_folder_cache:
+            light_folder_cache[lfolder] = os.listdir(lfolder)
+
+        # Extract base name and extension
+        lmatch = versioned_pat.match(original_lightname)
+        if not lmatch:
+            problems["row"].append(i)
+            problems["oldname"].append(original_lightname)
+            problems["newname"].append("Regex match failed")
+            continue
+
+        base = lmatch.group("base")
+        rev = int(lmatch.group("rev"))
+        ext = lmatch.group("ext")
+
+        # Find all matching versions in the folder
+        lmatches = []
+        for f in light_folder_cache[lfolder]:
+            m = versioned_pat.match(f)
+            if m:
+                f_base = m.group("base")
+                f_rev = int(m.group("rev"))
+                f_ext = m.group("ext")
+        
+                if f_base == base and f_ext == ext:
+                    lmatches.append((f, f_rev))
+        
+        if lmatches:
+            newest_file = max(lmatches, key=lambda x: x[1])[0]
+            if original_lightname != newest_file:
+                new_lightnames.loc[i] = newest_file
+                changes["row"].append(i)
+                changes["oldname"].append(original_lightname)
+                changes["newname"].append(newest_file)
+
+        # DARKS ------------------------------
+        if dfolder not in dark_folder_cache:
```

Comment:
dark handling code copy pasted from light handling code above is begging for encapsulation, but if it works it works.

Reply:
This function has been fullly rebuilt to address this in 43aca59752dfa8dcfe3b41ed84d20b0d3feb7087